### PR TITLE
AdamW refactor (current)

### LINKED ISSUE #104899: Refactor Adam and AdamW by abstracting out common code
AdamW differs with Adam only in the weight_decay handling. Everything else is the same. We should reuse code instead of hosting the same exact logic in two places.

One way to do this is to have certain functions in adam.py that can be accessed in adamw.py. The main work that would need to be done would be consolidating the forloop and foreach implementations. The fused implementation already takes advantage of the commonality. 

cc @vincentqb @jbschlosser @albanD @crcrpar

### PR DESCRIPTION
**Pull request by tfsingh**
Fixes #104899, reopening the PR as discussed.

cc @janeyx99

### CODE REVIEW

**Review by janeyx99 — DISMISSED** [review]
Thanks!

File: torch/optim/adam.py
Comment (janeyx99) [inline torch/optim/adam.py:1]: Oh actually, you need to add a line like
`            group.setdefault('decoupled_weight_decay', False)`
in Adam's `__setstate__`!

**Comment by huydhn** [comment]
I want to borrow your PR a bit to test bc linter, so please don't worry when I remove the `suppress-bc-linter` label.

**Comment by janeyx99** [comment]
oh shoot did I forget to merge this

**Pull request closed without merging**